File: app/src/main/java/workshop/akbolatss/tagsnews/screen/sources/SourcesAdapter.java
package workshop.akbolatss.tagsnews.screen.sources;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import workshop.akbolatss.tagsnews.R;
import workshop.akbolatss.tagsnews.model.dao.RssSource;
import workshop.akbolatss.tagsnews.screen.sources.helper.ItemTouchHelperAdapter;
import workshop.akbolatss.tagsnews.screen.sources.helper.ItemTouchHelperViewHolder;

public class SourcesAdapter extends RecyclerView.Adapter<SourcesAdapter.NewsHolder> implements ItemTouchHelperAdapter {

    private List<RssSource> mSourcesList;

    private final OnRssClickListener mClickListener;

    public SourcesAdapter(OnRssClickListener mClickListener) {
        mSourcesList = new ArrayList<>();
        this.mClickListener = mClickListener;
    }

    @Override
    public NewsHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater mLayoutInflater = LayoutInflater.from(parent.getContext());
        View view = mLayoutInflater.inflate(R.layout.rv_source_item, parent, false);
        return new NewsHolder(view);
    }

    @Override
    public void onBindViewHolder(NewsHolder holder, int position) {
        RssSource rssSource = mSourcesList.get(position);
        holder.bind(rssSource, mClickListener);

    }

    @Override
    public int getItemCount() {
        if (mSourcesList == null) {
            return 0;
        }
        return mSourcesList.size();
    }

    public void onAddItems(List<RssSource> rssItems) {
        if (rssItems != null) {
            mSourcesList.clear();
            mSourcesList.addAll(rssItems);
            notifyDataSetChanged();
        }
    }

    public void onAddItem(RssSource rssSource) {
        if (rssSource != null) {
            mSourcesList.add(rssSource);
            notifyItemInserted(mSourcesList.size() + 1);
        }
    }

    public void onUpdateItem(RssSource rssSource, int pos) {
        if (rssSource != null) {
            mSourcesList.set(pos, rssSource);
            notifyItemChanged(pos);
        }
    }

    public void onRemoveItem(int pos) {
        mSourcesList.remove(pos);
        notifyItemRemoved(pos);
    }

    @Override
    public boolean onItemMove(int fromPosition, int toPosition) {
        Collections.swap(mSourcesList, fromPosition, toPosition);
        notifyItemMoved(fromPosition, toPosition);
        return true;
    }

    @Override
    public void onItemsMoved(int fromPosition, int toPosition) {
        mClickListener.onItemsSwapped(mSourcesList.get(fromPosition), mSourcesList.get(toPosition));
    }

    public interface OnRssClickListener {

        public void onItemsSwapped(RssSource from, RssSource to);

        public void onSourceOptions(RssSource rssSource, View view, int pos);

        public void onSourceSwitch(RssSource rssSource, boolean isActive, int pos);
    }

    public class NewsHolder extends RecyclerView.ViewHolder implements ItemTouchHelperViewHolder {

//        @BindView(R.id.cardView)
//        CardView cardView;
//        @BindView(R.id.tvTitle)
//        TextView mTitle;
//        @BindView(R.id.tvLink)
//        TextView mLink;
//        @BindView(R.id.imgIcon)
//        ImageView mImgIcon;
//        @BindView(R.id.imgOptions)
//        ImageView mImgOptions;
//        @BindView(R.id.cbInit)
        CheckBox mCheckBox;
        Context context;

        public NewsHolder(View itemView) {
            super(itemView);
            context = itemView.getContext();
        }

        public void bind(final RssSource rssSource, final OnRssClickListener listener) {
//            mTitle.setText(rssSource.getTitle());
//            mLink.setText(rssSource.getLink());
//            mCheckBox.setChecked(rssSource.getIsActive());
//
//            Picasso.with(context)
//                    .load(rssSource.getVisualUrl())
//                    .placeholder(R.drawable.ic_rss_feed_24dp)
//                    .error(R.drawable.ic_rss_feed_24dp)
//                    .into(mImgIcon);
//
//            mImgOptions.setOnClickListener(new View.OnClickListener() {
//                @Override
//                public void onClick(View v) {
//                    listener.onSourceOptions(rssSource, v, getLayoutPosition());
//                }
//            });

            mCheckBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    listener.onSourceSwitch(rssSource, isChecked, getLayoutPosition());
                }
            });
        }

        @Override
        public void onItemSelected() {
//            cardView.setCardBackgroundColor(context.getResources().getColor(R.color.colorCardDragged));
        }

        @Override
        public void onItemClear() {
//            cardView.setCardBackgroundColor(context.getResources().getColor(R.color.colorCardBackground));
        }
    }
}

File: app/src/main/java/workshop/akbolatss/tagsnews/screen/sources/SourcesAdapter.kt
package workshop.akbolatss.tagsnews.screen.sources

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.rv_source_item.view.*
import workshop.akbolatss.tagsnews.R
import workshop.akbolatss.tagsnews.model.dao.RssSource
import workshop.akbolatss.tagsnews.screen.sources.helper.ItemTouchHelperAdapter
import workshop.akbolatss.tagsnews.screen.sources.helper.ItemTouchHelperViewHolder
import java.util.*

/**
 * Custom RecyclerView.Adapter for RSS source items
 */
class SourcesAdapter(private val mClickListener: SourceListener) : RecyclerView.Adapter<SourcesAdapter.NewsHolder>(), ItemTouchHelperAdapter {

    private val mSourcesList: MutableList<RssSource> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsHolder {
        val mLayoutInflater = LayoutInflater.from(parent.context)
        val view = mLayoutInflater.inflate(R.layout.rv_source_item, parent, false)
        return NewsHolder(view)
    }

    override fun onBindViewHolder(holder: NewsHolder, position: Int) {
        val rssSource = mSourcesList[position]
        holder.bind(rssSource, mClickListener)
    }

    override fun getItemCount(): Int {
        return mSourcesList.size
    }

    /**
     * Load list of items to this adapter
     */
    fun onAddItems(rssItems: List<RssSource>?) {
        if (rssItems != null) {
            mSourcesList.clear()
            mSourcesList.addAll(rssItems)
            notifyDataSetChanged()
        }
    }

    /**
     * Add item to this adapter
     */
    fun onAddItem(rssSource: RssSource?) {
        if (rssSource != null) {
            mSourcesList.add(rssSource)
            notifyItemInserted(mSourcesList.size + 1)
        }
    }

    /**
     * Update item in specific position
     * @param pos position
     */
    fun onUpdateItem(rssSource: RssSource?, pos: Int) {
        if (rssSource != null) {
            mSourcesList[pos] = rssSource
            notifyItemChanged(pos)
        }
    }

    /**
     * Remove item from specific position
     * @param pos position
     */
    fun onRemoveItem(pos: Int) {
        mSourcesList.removeAt(pos)
        notifyItemRemoved(pos)
    }

    /**
     * Swap items
     */
    override fun onItemMove(fromPosition: Int, toPosition: Int): Boolean {
        Collections.swap(mSourcesList, fromPosition, toPosition)
        notifyItemMoved(fromPosition, toPosition)
        return true
    }

    override fun onItemsMoved(fromPosition: Int, toPosition: Int) {
        mClickListener.onItemsSwapped(mSourcesList[fromPosition], mSourcesList[toPosition])
    }

    interface SourceListener {

        fun onStartDrag(viewHolder: RecyclerView.ViewHolder)

        fun onItemsSwapped(from: RssSource, to: RssSource)

        fun onSourceOptions(rssSource: RssSource, view: View, pos: Int)

        fun onSourceSwitch(rssSource: RssSource, isActive: Boolean, pos: Int)
    }

    /**
     * Custom View Holder where View instantiates
     */
    inner class NewsHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(rssSource: RssSource, listener: SourceListener) {
            itemView.tvTitle.text = rssSource.title
            itemView.tvLink.text = rssSource.link
            itemView.tvSubscribers.text = rssSource.subscribers.toString() + " Subscribers"
            itemView.tvDescription.text = rssSource.description

            Picasso.with(itemView.context)
                    .load(rssSource.visualUrl)
                    .placeholder(R.drawable.ic_rss_feed_24dp)
                    .error(R.drawable.ic_rss_feed_24dp)
                    .into(itemView.imgIcon)

            iconState(rssSource.isActive)

            itemView.imgOptions.setOnClickListener {
                listener.onSourceOptions(rssSource, it, adapterPosition)
            }

            itemView.imgAdd.setOnClickListener {
                listener.onSourceSwitch(rssSource, !rssSource.isActive, adapterPosition)
                iconState(rssSource.isActive)
            }
            itemView.flSwap.setOnTouchListener { _, _ ->
                listener.onStartDrag(this)
                return@setOnTouchListener true
            }
        }

        private fun iconState(isActive: Boolean) {
            if (isActive) {
                itemView.imgAdd.setImageResource(R.drawable.ic_done_24dp)
            } else {
                itemView.imgAdd.setImageResource(R.drawable.ic_add_24dp)
            }
        }
    }
}
